fuzzers/run_fuzzer.py

```diff
@@ -0,0 +1,528 @@
+#!/usr/bin/env python3
+
+from datetime import datetime, timedelta
+
+import io
+import os
+import os.path
+import re
+import resource
+import signal
+import subprocess
+import sys
+import traceback
+
+secs_in_min = 60
+secs_in_hour = 60 * secs_in_min
+
+
+def pretty_timedelta_str(d):
+    """Pretty print a time delta object.
+
+    >>> pretty_timedelta_str(timedelta(seconds=2))
+    '2s'
+
+    >>> pretty_timedelta_str(timedelta(seconds=61))
+    '1m01s'
+
+    >>> pretty_timedelta_str(timedelta(seconds=125))
+    '2m05s'
+
+    >>> pretty_timedelta_str(timedelta(seconds=secs_in_hour * 5))
+    '5h'
+
+    Once we get to displaying hours, we don't bother displaying seconds.
+    >>> pretty_timedelta_str(timedelta(seconds=secs_in_hour * 5 + 1))
+    '5h'
+
+    >>> pretty_timedelta_str(timedelta(
+    ...   seconds=secs_in_hour * 5 + secs_in_min * 2 + 1))
+    '5h02m'
+
+    >>> pretty_timedelta_str(timedelta(seconds=secs_in_hour * 5.5))
+    '5h30m'
+    """
+    assert isinstance(d, timedelta)
+    ts = d.total_seconds()
+    assert ts > 0, (d, ts)
+
+    bits = []
+    hours = int(ts / secs_in_hour)
+    if hours > 0:
+        bits.append("{}h".format(hours))
+        ts -= hours * secs_in_hour
+
+    mins = int(ts / secs_in_min)
+    if mins > 0:
+        if hours > 0:
+            bits.append("{:02d}m".format(mins))
+        else:
+            bits.append("{:d}m".format(mins))
+        ts -= mins * secs_in_min
+
+    if hours == 0:
+        if mins > 0:
+            bits.append("{:02d}s".format(int(ts)))
+        else:
+            bits.append("{:d}s".format(int(ts)))
+
+    assert len(bits) > 0, d
+
+    return "".join(bits)
+
+
+class Logger:
```

Comment:
Why have an additional output buffer at all?  print() is to a buffered writer?

Reply:
You can't use `print()` (or any io function) in a signal handler.

Reply:
And you can't call malloc either, so what's you point?  This code is doubly wrong then?

Reply:
Putting it on a non-preallocated queue doesn't solve the problem.  List,append can result in a memory allocation, which is not safe in a signal handler either.

Reply:
> A Python signal handler does not get executed inside the low-level (C) signal handler. Instead, the low-level signal handler sets a flag which tells the virtual machine to execute the corresponding Python signal handler at a later point (for example at the next bytecode instruction).

This means that memory allocation is fine in a Python signal handler. At first glance this also makes it seem like IO should be also be fine. However, there are a bunch of complicated reasons why it is very *occasionally* not.